bfvmm/src/vmcs/src/vmcs_intel_x64_debug.cpp

```diff
@@ -22,524 +23,716 @@
-    PRINT_FIELD(true, VMCS_HOST_IA32_SYSENTER_ESP);
-    PRINT_FIELD(true, VMCS_HOST_IA32_SYSENTER_EIP);
-    PRINT_FIELD(true, VMCS_HOST_RSP);
-    PRINT_FIELD(true, VMCS_HOST_RIP);
+    dump_vmcs_field(host_cr0::addr, host_cr0::name, host_cr0::exists());
+    dump_vmcs_field(host_cr3::addr, host_cr3::name, host_cr3::exists());
+    dump_vmcs_field(host_cr4::addr, host_cr4::name, host_cr4::exists());
+    dump_vmcs_field(host_fs_base::addr, host_fs_base::name, host_fs_base::exists());
+    dump_vmcs_field(host_gs_base::addr, host_gs_base::name, host_gs_base::exists());
+    dump_vmcs_field(host_tr_base::addr, host_tr_base::name, host_tr_base::exists());
+    dump_vmcs_field(host_gdtr_base::addr, host_gdtr_base::name, host_gdtr_base::exists());
+    dump_vmcs_field(host_idtr_base::addr, host_idtr_base::name, host_idtr_base::exists());
+    dump_vmcs_field(host_ia32_sysenter_esp::addr, host_ia32_sysenter_esp::name, host_ia32_sysenter_esp::exists());
+    dump_vmcs_field(host_ia32_sysenter_eip::addr, host_ia32_sysenter_eip::name, host_ia32_sysenter_eip::exists());
+    dump_vmcs_field(host_rsp::addr, host_rsp::name, host_rsp::exists());
+    dump_vmcs_field(host_rip::addr, host_rip::name, host_rip::exists());
 }
 
 void
-vmcs_intel_x64::print_execution_controls()
+debug::dump_vmx_controls()
 {
-    print_pin_based_vm_execution_controls();
-    print_primary_processor_based_vm_execution_controls();
-    print_secondary_processor_based_vm_execution_controls();
-    print_vm_exit_control_fields();
-    print_vm_entry_control_fields();
+    bfdebug << bfendl;
+    dump_pin_based_vm_execution_controls();
+    dump_primary_processor_based_vm_execution_controls();
+    dump_secondary_processor_based_vm_execution_controls();
+    dump_vm_exit_control_fields();
+    dump_vm_entry_control_fields();
 }
 
 void
-vmcs_intel_x64::print_pin_based_vm_execution_controls()
+debug::dump_pin_based_vm_execution_controls()
 {
-    auto controls =
-        vmread(VMCS_PIN_BASED_VM_EXECUTION_CONTROLS);
-
     bfdebug << "----------------------------------------" << bfendl;
     bfdebug << "- Pin VM Execution Controls            -" << bfendl;
     bfdebug << "----------------------------------------" << bfendl;
 
-    PRINT_CONTROL(VM_EXEC_PIN_BASED_EXTERNAL_INTERRUPT_EXITING);
-    PRINT_CONTROL(VM_EXEC_PIN_BASED_NMI_EXITING);
-    PRINT_CONTROL(VM_EXEC_PIN_BASED_VIRTUAL_NMIS);
-    PRINT_CONTROL(VM_EXEC_PIN_BASED_ACTIVATE_VMX_PREEMPTION_TIMER);
-    PRINT_CONTROL(VM_EXEC_PIN_BASED_PROCESS_POSTED_INTERRUPTS);
+    dump_vm_control(pin_based_vm_execution_controls::external_interrupt_exiting::name,
+                    pin_based_vm_execution_controls::external_interrupt_exiting::is_enabled(),
+                    ia32_vmx_true_pinbased_ctls::external_interrupt_exiting::is_allowed0(),
+                    ia32_vmx_true_pinbased_ctls::external_interrupt_exiting::is_allowed1());
+
+    dump_vm_control(pin_based_vm_execution_controls::nmi_exiting::name,
```

Comment:
Please move the implementation to the headers as inline functions and ditch the .cpp